Formatter/Pool.php

```diff
@@ -29,20 +31,60 @@ class Pool
     protected $defaultFormatter;
 
     /**
+     * NEXT_MAJOR: use LoggerAwareTrait.
+     *
      * @var LoggerInterface
      */
     protected $logger;
 
     /**
-     * @param LoggerInterface|null $logger
-     * @param string|null          $defaultFormatter
+     * @param string|null $defaultFormatter
      */
-    public function __construct(LoggerInterface $logger = null, $defaultFormatter = null)
+    public function __construct($defaultFormatter = null)
     {
-        $this->logger = $logger;
+        if (func_num_args() == 2) {
+            list($logger, $defaultFormatter) = func_get_args();
+            $this->logger = $logger;
+            $this->defaultFormatter = $defaultFormatter;
+        } elseif (func_num_args() == 1) {
+            if ($defaultFormatter instanceof LoggerInterface) {
+                $this->logger = $defaultFormatter;
+            } else {
+                // NEXT_MAJOR: Only keep this block
+                $this->defaultFormatter = $defaultFormatter;
+            }
+        }
 
-        // TODO: This should become a required first parameter when the major version is changed
-        $this->defaultFormatter = $defaultFormatter;
+        // NEXT_MAJOR: keep the else block only
+        if ($this->logger) {
+            @trigger_error(sprintf(
+                'Providing a logger to %s through the constructor is deprecated since 3.x'.
```

Comment:
This is where the string split kicks in as well: if you change the message, you need to rearrange the strings as well.